implement snapshot for a thindev

trgill opened the pull request:
Taking a snapshot of a ThinDev creates a new ThinDev that is no different
than any other thin provisioned device.

Documentation for snapshots can be found at:

https://www.kernel.org/doc/Documentation/device-mapper/thin-provisioning.txt

See section i) Creating an internal snapshot.

trgill commented:
@mulkieran @agrover review please?

agrover commented:
Why 5a977ce? Device is `Copy` so it shouldn't be necessary, is my thought.

trgill commented:
@agrover you are right.  I removed the commit.

mulkieran commented on src/thindev.rs:
```diff
@@ -184,6 +184,36 @@ impl ThinDev {
     }
 
+    /// Create a snapshot of a ThinDev.  Once created a snapshot
+    /// is the same as any other thin provisioned device.  There is
```
Does it make more sense to define this method on a ThinPoolDev? Like:

```
pub fn snapshot(&self, dm: &DM, origin: &DevId, origin_thin_id: ThinDevId, snapshot_name: &DmName, snapshot_thin_id: ThinDevId) {
  dm.device_suspend(origin, ...);
  self.message(dm, &format(..., snapshot_thin_id, origin_thin_id));
  dm.device_suspend(origin, ...);
  let dev_info = device_create(dm, snapshot_name, &ThinDev::dm_table(self.device(), snapshot_thin_id, <maybe size would have to be an addtional param>);
  Ok(...)
}
```

trgill replied:
I was thinking snapshot() is similar to ThinDev::new() and ThinDev::setup() since it returns a new ThinDev.

Why are you thinking it belongs in ThinPoolDev?  Because we message the ThinPoolDev?

agrover replied:
honestly I could see it either way...

mulkieran replied:
There is at least one significant difference. In ThinPoolDev snapshot method it is necessary to pass the size of the snapshot. But I think that for a (theoretically general purpose) library, that might be a good thing. What happens if snapshot size > origin size, or vice-versa?

Also, parameters are a bit less odd, in my version, I think.

trgill replied:
Snapshot size is always equal to the source size - they are identical at the time of the snapshot.

I'll implement what @mulkieran suggested and see how it works.  Will report back.

mulkieran replied:
There are two things that happen in the method:

1. The thin pool gets informed of a snapshot, and gets the relation via two thin ids, one the snapshot's and one the origins.
2. The snapshot device is created and the size is specified. This is just another thin device with the snapshot id and a specified size. What would happen if you specified a larger size at this time? Would table_load fail? Would it succeed, and if it did, what would be the consequences?

trgill commented:
So, this doesn't compile if I move it to thinpooldev.rs - the fields are private.

```
        Ok(ThinDev {
               dev_info: dev_info,
               thin_id: snapshot_thin_id,
               size: self.size(),
               thinpool: self.device(),
           })
```

I'll add a new method to thindev.rs to construct a ThinDev given these fields?

trgill commented:
@mulkieran yeah, I've got a test in a stratisd PR.

mulkieran commented:
@trgill I wasn't clear. In devicemapper-rs, there is now an existing loopbacked based testing infrastructure. In both thinpool.rs and thinpooldev.rs there are testing modules w/ some tests already written. You should be able to throw in an additional test for this new function, regardless of where it ends up, either as part of ThinDev or as part of ThinPoolDev.

mulkieran commented:
@trgill Maybe you are better off putting it in ThinPool. But you could still change the parameters as:

```
pub fn snapshot(&self,
                           dm: &DM,
                           thin_pool: &ThinPoolDev,
                           snapshot_name: &DmName,
                           snapshot_thin_id: ThinDevId,
                           size: Sectors) { // size might be optional, defaulting to self.size() ?
  let source_id = DevId::Name(self.name());
  dm.device_suspend(&source_id, ...);
  thin_pool.message(dm, &format(...));
  dm.device_suspend(&source_id, ...);
  let table = ThinDev::dm_table(thin_pool.device(), snapshot_thin_id, size);
  let dev_info = device_create(dm, snapshot_name, &table);
  Ok(...)
}
```
Seems to bring it into line a bit more w/ other ThinDev methods.

trgill commented:
@mulkieran I updated the parameters to ThinDev::snapshot() as you suggested.

I also added a very simple unit test that just compares the snapshot size with the source.  I could enhance to write some files - but my plan was to do that as an integration test.

mulkieran commented on src/thindev.rs:
```diff
@@ -184,6 +184,35 @@ impl ThinDev {
+                    -> DmResult<ThinDev> {
+        let source_name = DevId::Name(self.name());
+        dm.device_suspend(&source_name, DM_SUSPEND)?;
+        dm.target_msg(&DevId::Name(thin_pool.name()),
```
You can use ```thin_pool.messag(...``` here.

trgill replied:
Ok

mulkieran commented on src/thindev.rs:
```diff
@@ -184,6 +184,35 @@ impl ThinDev {
+                    snapshot_name: &DmName,
+                    snapshot_thin_id: ThinDevId)
+                    -> DmResult<ThinDev> {
+        let source_name = DevId::Name(self.name());
```
Minor, but this really is an id, not a name.

trgill replied:
Ok

trgill commented:
@mulkieran 
@agrover maybe you know the answer?

I don't think I said a snapshot couldn't be made larger or smaller than the origin.  The documentation shows the snapshot table being the same as the origin.

The following is from linux/drivers/md/dm-thin.c at ~line 53

 ```
* When you take an internal snapshot you clone the root node
 * of the origin btree.  After this there is no concept of an origin or a
 * snapshot.  They are just two device trees that happen to point to the
 * same data blocks.

```

How about I open an issue to investigate?  I don't think we want to hold this PR.

trgill commented:
I opened issue #177 to track concerns about changing size of snapshot table.

trgill commented:
rebased

agrover commented:
👍
